lib/OpenQA/Worker.pm

```diff
@@ -611,36 +616,46 @@ sub is_ovs_dbus_service_running ($self) {
     catch ($e) { return 0 }
 }
 
-# checks whether the worker is available
+# returns whether the worker is available and a reason
 # note: This is used to check certain error conditions *before* starting a job to prevent incompletes and
 #       being able to propagate the brokenness to the web UIs.
+# note: High load will yield a corresponding error message as reason so the worker becomes broken and
+#       thus will not pick up any new jobs. However, a worker under high load is still considered being
+#       able to work on jobs that are already enqueued. Hence this function returns a "1" for the
+#       availability in this case.
 sub check_availability ($self) {
     # check whether the cache service is available if caching enabled
     if (my $cache_service_client = $self->{_cache_service_client}) {
         my $error = $cache_service_client->info->availability_error;
         my $host = $cache_service_client->host // '?';
-        return "Worker cache not available via $host: $error" if $error;
+        return (0, "Worker cache not available via $host: $error") if $error;
     }
 
     # check whether qemu is still running
     if (my $qemu_pid = $self->is_qemu_running) {
-        return "A QEMU instance using the current pool directory is still running (PID: $qemu_pid)";
+        return (0, "A QEMU instance using the current pool directory is still running (PID: $qemu_pid)");
     }
 
-    # avoid running jobs if system utilization is critical and ensure pool directory is locked
-    if (my $error = $self->_check_system_utilization || $self->_setup_pool_directory) {
-        return $error;
-    }
+    # ensure pool directory is locked
+    if (my $error = $self->_setup_pool_directory) { return (0, $error) }
 
     # auto-detect worker address if not specified explicitly
     my $settings = $self->settings;
-    return 'Unable to determine worker address (WORKER_HOSTNAME)' unless $settings->auto_detect_worker_address;
+    return (0, 'Unable to determine worker address (WORKER_HOSTNAME)') unless $settings->auto_detect_worker_address;
 
     # check org.opensuse.os_autoinst.switch if it is a MM-capable worker slot
-    return "D-Bus service '$self->{_ovs_dbus_service_name}' is not running"
+    return (0, "D-Bus service '$self->{_ovs_dbus_service_name}' is not running")
       if $settings->has_class('tap') && !$self->is_ovs_dbus_service_running;
 
-    return undef;
+    # avoid picking up new jobs if system utilization is critical
+    return (1, $self->_check_system_utilization);
```

Comment:
this sounds a bit confusing to me. It says "avoid picking up new jobs" but it is returning a 1 which means success, right?

Reply:
The "1" is so it still continues with the execution of already enqueued jobs. It will just avoid picking up new jobs and that's the 2nd return value. I changed the phrasing. This should be good enough considering I also added general explanation at the top of the function.